cnd/src/swap_protocols/rfc003/ledger_state.rs

```diff
@@ -18,26 +25,136 @@ pub enum LedgerState<L: Ledger> {
         htlc_location: L::HtlcLocation,
         deploy_transaction: L::Transaction,
         fund_transaction: L::Transaction,
+        asset: A,
     },
     Redeemed {
         htlc_location: L::HtlcLocation,
         deploy_transaction: L::Transaction,
         fund_transaction: L::Transaction,
         redeem_transaction: L::Transaction,
+        asset: A,
+        secret: Secret,
     },
     Refunded {
         htlc_location: L::HtlcLocation,
         deploy_transaction: L::Transaction,
         fund_transaction: L::Transaction,
         refund_transaction: L::Transaction,
+        asset: A,
     },
     IncorrectlyFunded {
         htlc_location: L::HtlcLocation,
         deploy_transaction: L::Transaction,
         fund_transaction: L::Transaction,
+        asset: A,
     },
 }
 
+impl<L: Ledger, A: Asset> LedgerState<L, A> {
+    pub fn transition_to_deployed(&mut self, deployed: Deployed<L>) {
+        let Deployed {
+            transaction,
+            location,
+        } = deployed;
+
+        match std::mem::replace(self, LedgerState::NotDeployed) {
+            LedgerState::NotDeployed => {
+                *self = LedgerState::Deployed {
+                    deploy_transaction: transaction,
+                    htlc_location: location,
+                }
+            }
+            other => panic!("expected state NotDeployed, got {}", HtlcState::from(other)),
+        }
+    }
+
+    pub fn transition_to_funded(&mut self, funded: Funded<L, A>) {
+        let Funded { transaction, asset } = funded;
+
+        match std::mem::replace(self, LedgerState::NotDeployed) {
+            LedgerState::Deployed {
+                deploy_transaction,
+                htlc_location,
+            } => {
+                *self = LedgerState::Funded {
+                    deploy_transaction,
+                    htlc_location,
+                    fund_transaction: transaction,
+                    asset,
+                }
+            }
+            other => panic!("expected state Deployed, got {}", HtlcState::from(other)),
+        }
+    }
+
+    pub fn transition_to_incorrectly_funded(&mut self, funded: Funded<L, A>) {
+        let Funded { transaction, asset } = funded;
+
+        match std::mem::replace(self, LedgerState::NotDeployed) {
```

Comment:
:question: Why are the `mem::replace`s needed and why are they always replaced with `LedgerState::NotDeployed`?

Reply:
I wondered too but it then made sense ;)